### ISSUE TITLE
git-stree vs. git subrepo

### ISSUE THREAD

**Issue by JaKXz** [deliciousinsights/git-stree#20]
Hi! This looks awesome. 

However, I was browsing Homebrew's formulas and stumbled on [git subrepo](https://github.com/ingydotnet/git-subrepo/). At the surface it looks like you guys are solving the same problem, so I'm just curious if there are major differences to be considered. I'll write back if I find anything but if you know off the top of your head that would be great.

Thanks!

**Comment by gvt** [comment]
I think this is a good question. I am seeking a solution for the use case git-stree solves and git-subrepo does seem to solve the same issue. I like that git-subrepo leaves a `.gitrepo` file but dislike that its not built on top of a git-standard thing such as subtrees (if one could say that subtrees is now an "official" git thing). I like that git-strees builds on top of git's subtree stuff but really would want clones of the containing repo to be aware that it contains git-strees from the start.

**Comment by tdd** [comment]
Hey,

That's very much what #3 is about. Unlike what submodules do, `git stree` would auto-detect these things and auto-adjust based on them. That might hit the sweet spot you're looking for :smile:

**Comment by tdd** [comment]
btw guys, this has already been discussed in #18 :wink:

**Comment by gvt** [comment]
Oh yes I see, thanks. Hope you and Ingy have a chat!

**Comment by ingydotnet** [comment]
@gvt I wanted to weigh in on your comment about building over subtree. The
git-subrepo project got its start by looking over the subtree codebase,
realizing that it had good ideas but could be done much better, and then doing
it much better.

The subtree command is just a bash script that uses other git plumbing commands
to do its work. The subrepo command is the same thing. Building one script over
the other wouldn't really gain anything.

The subtree command started off as a github repo:
https://github.com/apenwarr/git-subtree/ and then was moved into the git core,
but only as a 3rd party contribution:
https://github.com/git/git/tree/master/contrib

So while it is now maintained as part or core git, it is not quite at the same
level as the other core commands. What I assume this means is that the core
maintainers thought it a good enough idea to ship to everyone, but has not
embraced it as a final solution.

More importantly, I wanted to point out that I am ideologically adverse to
subrepo being a core git command. The reason for this is twofold:

First off, the only people who need subrepo installed are the people who are
cloning subrepos and/or push pulling them. This is really only a project owner
and *maybe* a collaborator. The people using the repos (with subrepos) don't
need the command or even need to know about it.

Secondly, the subrepo command isn't tied to any specific version of git, and so
making it core would just mean that people using older gits wouldn't have the
latest subrepo. In fact, making it a wrapper around subtree actually would make
it deal with various version specific issues of subtree.

In other words, the audience of people using `git-subrepo` is quite small (as
compared to those needing `git-submodule`) and so those people should always
have the latest and most up to date version, regardless of the git version they
are on.

Hope this helps, @ingydotnet

**Comment by stela5** [comment]
`git-subrepo` looks promising but it's difficult to get working on Windows: https://github.com/ingydotnet/git-subrepo/issues/38

**Comment by tdd** [comment]
Hey @ingydotnet as I finally can spare some time to look at git-stree again and decide whether to do a complete rewrite (which I was about to do) or whether to call it a day and encourage people to use git-subrepo, I looked more in-depth into git-subrepo usage.

Generally, this is way more advanced than stree. I like the level of UI/UX (smart outputs, smart storage, proper docs and man page, proper tests, good dual-shell completion…). However, before I decide whether to stop my own efforts, there are a couple things I worry about. Would you care to spend a moment answering these questions? I'd appreciate it.

1. [Apparently](https://github.com/ingydotnet/git-subrepo/issues/38) it doesn't work yet in Git for Windows's Git Bash. It's been updated recently to a much more recent Bash and Git. Is that scheduled soon?
2. I don't see any way to push only selected local work back upstream. I get that request many times, and this is one of the core tenets of stree. What about subrepo?
3. The README speaks of various things nowhere to be found: the `…/upstream` ref mentioned everywhere is apparently `…/fetch` now, and the `git subrepo upgrade` feature doesn't seem to exist. What gives?

These are my main concerns. Everything else looks dandy, I must say.

Thank you for your time,

**Comment by ingydotnet** [comment]
Hi Christophe,

Just saw this. I think the best way to approach this is to move this email
to GitHub issues so that other people working on subrepo can chine in.

I've gone ahead and done that:
https://github.com/ingydotnet/git-subrepo/issues/116

Let's see what happens.

I'd love to join forces. Many thanks for reaching out.

Ingy

**Comment by JaKXz** [comment]
:+1: yay :)

**Comment by tdd** [comment]
Hey Ingy,

Well, this **was** on GitHub issues, to start with :wink:


On Mon, Jan 4, 2016, at 19:11, Ingy döt Net wrote:
> Hi Christophe,
>
>
Just saw this. I think the best way to approach this is to move
this email
>
to GitHub issues so that other people working on subrepo can chine in.
>
>
I've gone ahead and done that:
>
https://github.com/ingydotnet/git-subrepo/issues/116
>
>
Let's see what happens.
>
>
I'd love to join forces. Many thanks for reaching out.
>
>
Ingy
>
>
>
On Fri, Jan 1, 2016 at 3:14 AM, Christophe Porteneuve <
>
notifications@github.com> wrote:
>
>
> Hey @ingydotnet <https://github.com/ingydotnet> as I finally can spare
>
> some time to look at git-stree again and decide whether to do a
> complete
>
> rewrite (which I was about to do) or whether to call it a day and
> encourage
>
> people to use git-subrepo, I looked more in-depth into git-
> subrepo usage.
>
>
>
> Generally, this is way more advanced than stree. I like the level
> of UI/UX
>
> (smart outputs, smart storage, proper docs and man page, proper
> tests, good
>
> dual-shell completion…). However, before I decide whether to
> stop my own
>
> efforts, there are a couple things I worry about. Would you care
> to spend a
>
> moment answering these questions? I'd appreciate it.
>
>
>
>    1. Apparently <https://github.com/ingydotnet/git-subrepo/issues/38>
>       it
>
>    doesn't work yet in Git for Windows's Git Bash. It's been updated
>    recently
>
>    to a much more recent Bash and Git. Is that scheduled soon?
>
>    2. I don't see any way to push only selected local work back
>       upstream.
>
>    I get that request many times, and this is one of the core tenets
>    of stree.
>
>    What about subrepo?
>
>    3. The README speaks of various things nowhere to be found: the
>
>    …/upstream ref mentioned everywhere is apparently …/fetch now, and
>    the git
>
>    subrepo upgrade feature doesn't seem to exist. What gives?
>
>
>
> These are my main concerns. Everything else looks dandy, I must say.
>
>
>
> Thank you for your time,
>
>
>
> —
>
> Reply to this email directly or view it on GitHub
>
> <https://github.com/deliciousinsights/git-stree/issues/20#issuecomment-168302200>
>
> .
>
>
> — Reply to this email directly or view it on GitHub[1].


>

--
Christophe Porteneuve
tdd@tddsworld.com
 


Links:

  1. https://github.com/deliciousinsights/git-stree/issues/20#issuecomment-168755738

**Comment by tdd** [comment]
Closing this in favor of ingydotnet/git-subrepo#116
